text/src/autogluon/text/automm/predictor.py

```diff
@@ -452,78 +435,153 @@ def fit(
+            # reload the predictor metadata
+            predictor = AutoMMPredictor._load_metadata(predictor=self, path=best_trial_path)
+            # construct the model
+            model = create_model(
+                config=predictor._config,
+                num_classes=predictor._output_shape,
+                num_numerical_columns=len(predictor._df_preprocessor.numerical_feature_names),
+                num_categories=predictor._df_preprocessor.categorical_num_categories,
+                pretrained=False,  # set "pretrain=False" to prevent downloading online models
+            )
+            predictor._model = model
+            # average checkpoint
+            checkpoints_paths_and_scores = dict(
+                (os.path.join(checkpoint, RAY_TUNE_CHECKPOINT), score)
+                for checkpoint, score in analysis.get_trial_checkpoints_paths(best_trial, metric=metric)
+            )
+            # write checkpoint paths and scores to yaml file so that top_k_average could read it
+            best_k_model_path = os.path.join(best_trial_path, BEST_K_MODELS_FILE)
+            with open(best_k_model_path, "w") as yaml_file:
+                yaml.dump(checkpoints_paths_and_scores, yaml_file, default_flow_style=False)
+
+            last_ckpt_path = analysis.get_last_checkpoint(best_trial)
+            predictor._top_k_average(
+                model=predictor._model,
+                save_path=best_trial_path,
+                last_ckpt_path=last_ckpt_path,
+                minmax_mode=mode,
+                is_distill=is_distill,
+                top_k_average_method=predictor._config.optimization.top_k_average_method,
+                val_df=_fit_args["val_df"],
+                validation_metric_name=predictor._validation_metric_name,
+            )
+            cleanup_checkpoints(best_trial_path)
+            # move trial predictor one level up
+            contents = os.listdir(best_trial_path)
+            for content in contents:
+                shutil.move(
+                    os.path.join(best_trial_path, content),
+                    os.path.join(save_path, content),
+                )
+            shutil.rmtree(best_trial_path)
+            predictor._save_path = save_path
+
+            return predictor
+
+    def _hpo_fit_wrapper(self, sampled_hyperparameters, checkpoint_dir=None, **_fit_args):
+        from ray import tune
+
+        _fit_args[
+            "hyperparameters"
+        ] = sampled_hyperparameters  # The original hyperparameters is the search space, replace it with the hyperparameters sampled
+        _fit_args["save_path"] = tune.get_trial_dir()  # We want to save each trial to a separate directory
+        logger.debug(f"hpo trial save_path: {_fit_args['save_path']}")
+        if checkpoint_dir is not None:
+            _fit_args["resume"] = True
+            _fit_args["ckpt_path"] = os.path.join(checkpoint_dir, RAY_TUNE_CHECKPOINT)
+        self._fit(**_fit_args)
+
     def _setup_distillation(
```

Comment:
this method is too long. Can we split it into smaller logical units of work so it is easier to read and maintain?

Reply:
I think you refer to `_hyperparameter_tune()`This method is long because of many function signatures, for example `run()` and `_top_k_avetage()`, moving other things in to function will only short it by a little. For better readability, I've added comment on each logical operation.